### DISCUSSION TITLE
Flashing BMP to Nucleo STLink v2.1

### DISCUSSION THREAD

**Discussion by luuk00000000000000000000000000000000000** [blackmagic-debug/blackmagic#1514] [Q&A] [answered]
Hello,

I have a nucleo devboard with a built in STLink v2.1. I'd like to flash BMP to it so I can use it's STLink for general SWD purposes.

I however can't for the life of me seem to figure out how I can flash BMP to the STLink. 
I ran stlink-tool but it told me that the stlink wasn't in DFU mode, so I unplugged and replugged my board, and it didn't do anything.

If someone could point me in the right direction, that'd be great!

**Comment by dragonmux** [comment]
So.. your query brings a whole pile of questions:

1) What OS are you on?
2) What repository did you acquire stlink-tool from?
3) What version of the ST firmware is your ST-Link v2.1 running?
4) Have you tried using the [ST upgrader tool](https://www.st.com/en/development-tools/stsw-link007.html#get-software) to jump the board into its bootloader mode?
5) what version of the BMD firmware are you trying to use?

If you could answer some of those and try out the linked tool (it'll help you answer 3 as well), that would help to answer your query. Hope that helps some.

**Comment by luuk00000000000000000000000000000000000** [comment]
1. Windows 10 x64
2. https://github.com/sakana280/stlink-tool
3. V2J42M27
4. I think so? When I click the "open in update mode" button, I hear the windows device sound. However, I still can't upload anything with stlink-tool
5. The latest release would be fine.

I've also found out that the stlink resets when I short the reset pad at the bottom (duh), and then I get this output in stlink-tool:
```
Firmware version : V2J42S27
Loader version : 14152
ST-Link ID : 48FF6B064851705352431267
Firmware encryption key : [redacted]
Current mode : 2
ST-Link dongle is not in the correct mode. Please unplug and plug the dongle again.
```
Replugging does nothing.

**Comment by dragonmux** [comment]
We don't know the providence of that stlink-tool repository, however, we run our own with patches to fix compilation under modern GCC and which should work on Windows (using MSYS2, which is what the tool is intended to be built under): https://github.com/dragonmux/stlink-tool/tree/feature/compilation-fixes

Uwe's, on which ours is based, contains fixes for how the tool talks with ST-Link v2.1's and v3's.

**Comment by projectgus** [comment]
Can confirm dragonmux's advice is correct (thanks dragonmux!). I had an older version of stlink-tool from jeanthom's fork (not sure which commit) , and it fails in the same way on an old-ish STLinkV2.1 on a Nucleo:

```
❯ stlink-tool src/blackmagic.bin
Firmware version : V2J20S4
Loader version : 14152
ST-Link ID : 52FF6F065080565018280587
Firmware encryption key : 7E53534E9222CAFAFA00AF4D2464CCDC
Current mode : 2
ST-Link dongle is not in the correct mode. Please unplug and plug the dongle again.
```

The linked fork (commit 9ef778222) works fine:

```
❯ stlink-tool
Trying to switch BMP/Application to bootloader
StlinkV21 Bootloader found
Firmware version : V2J20S4
Loader version : 14152
ST-Link ID : 066FFF525056805087052818
Firmware encryption key : 7E53534E9222CAFAFA00AF4D2464CCDC
Current mode : 2
```

**Comment by dragonmux** [comment]
We have since formalised that repo properly as https://github.com/blackmagic-debug/stlink-tool - but this is not practically different to using the feature branch. We just wanted to make using the fork easier and particularly as the upstreams don't appear to be active any more.

**Comment by dragonmux** [comment, marked as answer]
Just a small update on this - the BMD fork of `stlink-tool` now builds and runs natively on Windows without any mucking about (via MSYS2 UCRT64, though in theory it can probably be built with MSVC now). This should now all just work for you if you use the latest `main` of the tool.

**Reply by luuk00000000000000000000000000000000000** [reply]
i haven't looked at this for a while but thank you for still answering after so long! 
i think i'll mark this as answered.
